core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java

```diff
@@ -115,7 +115,12 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
 
   @Override
   public InternalMessage.Builder payload(Object payload) {
-    this.payload = payload;
+    if (payload instanceof TypedValue<?>) {
+      this.payload = ((TypedValue) payload).getValue();
+      this.dataType = ((TypedValue) payload).getDataType();
```

Comment:
Internally the builder should really be using typedValue as state, such that if we add a field to TypedValue we don't need to edit this builder in order to ensure it gets propagated.